pkg/metrics/workqueue/naming.go

```diff
@@ -0,0 +1,71 @@
+package workqueue
+
+import (
+	"fmt"
+	"sync"
+)
+
+var (
+	nameProvidersInstance *nameProviders = &nameProviders{}
+)
+
+// NameProvider is the interface for name providers contributed by
+// packages that make use of k8s.io/client-go/util/workqueue.
+type NameProvider interface {
+	/*
+		Returns the subsystem name to be used for Prometheus workqueue metrics for
+		the given queue name. If the given queue name is not identified to belong to
+		the name provider's package, return value `ok` must be `false`.
+	*/
+	GetSubsystemFor(queueName string) (subsystem string, ok bool)
+}
+
+// NameProviderFunc is an adapter that allows to use functions with the right
+// signature as NameProvider.
+type NameProviderFunc func(queueName string) (subsystem string, ok bool)
+
+// GetSubsystemFor implements interface NameProvider.
+func (f NameProviderFunc) GetSubsystemFor(queueName string) (string, bool) {
+	return f(queueName)
+}
+
+// nameProviders is a composite name provider that queries all registered
+// name provides in the order of registration. The first positive result
+// will be returned.
+type nameProviders struct {
+	lock     sync.Mutex
+	children []NameProvider
```

Comment:
Why are you not using a map with queue name as key?

Reply:
First, this share package should not know it's consumers. This is why the naming must come from the packages where the respective queue belongs to.

Second, a name provider may also use pattern matching. A static name mapping would not allow that. Having a simple interface that is queried for each name looks reasonable to me.